### ISSUE TITLE
Some uncertainty about the demo

### ISSUE THREAD

**Issue by fatterZhang** [ChunhuanLin/deform_conv_pytorch#3]
First, thank you for your great job.
Then, as told in the paper, if the number of input feature map is N, the offset map should be 2N. But i found in your demo code as below:
`
class DeformNet(nn.Module):
    def __init__(self):
        super(DeformNet, self).__init__()
        self.conv1 = nn.Conv2d(1, 32, kernel_size=3, padding=1)
        self.bn1 = nn.BatchNorm2d(32)

        self.conv2 = nn.Conv2d(32, 64, kernel_size=3, padding=1)
        self.bn2 = nn.BatchNorm2d(64)

        self.conv3 = nn.Conv2d(64, 128, kernel_size=3, padding=1)
        self.bn3 = nn.BatchNorm2d(128)

        self.offsets = nn.Conv2d(128, 18, kernel_size=3, padding=1)
        self.conv4 = DeformConv2D(128, 128, kernel_size=3, padding=1)
        self.bn4 = nn.BatchNorm2d(128)

        self.classifier = nn.Linear(128, 10)

    def forward(self, x):
        # convs
        x = F.relu(self.conv1(x))
        x = self.bn1(x)
        x = F.relu(self.conv2(x))
        x = self.bn2(x)
        x = F.relu(self.conv3(x))
        x = self.bn3(x)

        # deformable convolution
        offsets = self.offsets(x)
        x = F.relu(self.conv4(x, offsets))
        x = self.bn4(x)

        x = F.avg_pool2d(x, kernel_size=28, stride=1).view(x.size(0), -1)
        x = self.classifier(x)

        return F.log_softmax(x, dim=1)
`
Does it shows that 18 offset maps and 128 input feature maps for deform_conv use ? 
And is it right?

**Comment by forrestl111** [comment]
As mentioned in the second chapter of the article, ***Both deformable convolution and RoI pooling modules operate on the 2D spatial domain***.  And ***N = |R|*** is also mentioned in the article, thus, N equals to the size of the kernel.

**Comment by bkvie** [comment]
My understanding is that, offsets are encoded in the channel dimension (e.g. conv. layer of n channels is paired with offset conv. layer of 2n channels)

**Comment by forrestl111** [comment]
In my opinion,  all the channels share the same offsets, but conv kernels applied to different locations of the feature map have different offsets. More details is showed in the *Section 2.1* of the article.

**Comment by John1231983** [comment]
@ChunhuanLin : I want to port the MXNet code as [here](https://github.com/msracver/Deformable-ConvNets/blob/master/fpn/symbols/resnet_v1_101_fpn_dcn_rcnn.py#L172) to your code

```
res3b3_branch2b_offset = mx.symbol.Convolution(name='res3b3_branch2b_offset', data=res3b3_branch2a_relu,
                                                           num_filter=72, pad=(1, 1), kernel=(3, 3), stride=(1, 1))
res3b3_branch2b = mx.contrib.symbol.DeformableConvolution(name='res3b3_branch2b', data=res3b3_branch2a_relu,
                                                           offset=res3b3_branch2b_offset,
                                                           num_filter=128, pad=(1, 1), kernel=(3, 3),
                                                           num_deformable_group=4,
                                                           stride=(1, 1), no_bias=True)
```
I found that when I set `num_filter=72`, the code does not work. It only work when I set the `num_filter=18`. Any suggestion? Thanks

**Comment by ChunhuanLin** [comment]
@John1231983 The output channels of the learned offsets should equal to 2*N, where N is the kernel size of the DeformableConvolution, in your case, N=3*3=9. So *num_filter* of *mx.symbol.Convolution* should be set to 18 to match the kernel size (3, 3) set in *mx.contrib.symbol.DeformableConvolution*.

**Comment by forrestl111** [comment]
@John1231983 In your case, *num_deformable_group* is set to 4 in the *DeformableConvolution* module, but unfortunately, group operation is not supported in my code yet. But it maybe work by adding parameter *[groups](http://pytorch.org/docs/0.3.1/nn.html#conv2d)* and set it to 4 [here](https://github.com/ChunhuanLin/deform_conv_pytorch/blob/master/deform_conv.py#L13).
